### DISCUSSION TITLE
TS error passing control as prop, only when context was passed to useForm

### DISCUSSION THREAD

**Discussion by RwwL** [react-hook-form/react-hook-form#6454] [Q&A]
This came up when trying to upgrade from v7.12.2 to v7.14.2. Not sure if it's really a bug vs. a usage error that's been uncovered by the upgrade so asking here first before filing an issue. 

My useForm call looks like, it includes that `shouldDisableUserDetails` bool passed as context which affects the yup validation based on specific conditions:
```
  const useFormMethods = useForm({
    context: { shouldDisableUserDetails },
    resolver: yupResolver(getValidationRules(t)),
  });
  const { control, handleSubmit } = useFormMethods;
 ```
 
Then later in the component I pass `control` to a form item wrapper that we use to wrap controlled components from a custom UI library (abridged example for clarity; I think all that matters is shown here):
 ```
<FormItemWrapper
  control={control}
  name="phone"
>
  <InputPhone />
</FormItemWrapper>
````

The TS compiler shows an error on `control=` there:
```
Type 'Control<FieldValues, { shouldDisableUserDetails: boolean; }>' is not assignable to type 'Control<FieldValues, object>'.
```

The relevant props for that `FormItemWrapper` look like this:
```
type Props<T> = {
  children: ReactNode;
  control: Control<T>;
  name: FieldPath<T>;
  rules?: RegisterOptions<T>;
};
```
And the start of the component looks like 
```
const FormItemWrapper = <T,>({
  children,
  control, // returned by the RHF useForm call in a parent
  name,
  rules, // define on <FormItem> usage and pass through to useController
}: Props<T>): JSX.Element => {
const {
    field,
    fieldState: { invalid, error },
  } = useController({
    name,
    control,
    rules: rules || {},
  });
  // etc etc
};
```
If I comment out the context, the compiler issues disappear. Again, this was all working fine in 7.12.2. FWIW, the setup with the generic types in the wrapper component 

Same issue whether I'm using TypeScript version v4.2.4 or v4.4.2, if that matters.

**Comment by julienfouilhe** [comment]
`useController` does not infer types from the `control` prop you pass to it unfortunately.
You need to pass them to it this way I think

```typescript
const FormItemWrapper = <TFieldValues extends FieldValues, TContext extends object>({
  children,
  control, // returned by the RHF useForm call in a parent
  name,
  rules, // define on <FormItem> usage and pass through to useController
}: Props<TFieldValues, TContext>): JSX.Element => {
const {
    field,
    fieldState: { invalid, error },
  } = useController<TFieldValues, TContext>({
    name,
    control,
    rules: rules || {},
  });
  // etc etc
};
```

**Reply by RwwL** [reply]
Thanks, @julienfouilhe, this is helpful and seems to get me closer but when I set it up this way the compiler complains about `TContext` on this line of your example:

`} = useController<TFieldValues, TContext>({`

The error message is:

```
Type 'TContext' does not satisfy the constraint 'Path<TFieldValues>'.
  Type 'Record<string, unknown>' is not assignable to type 'Path<TFieldValues>'.
```

(oh, and on the first line of your example, I changed `object` to `Record<string, unknown>` based on a warning from the compiler about that. Using `object` just leads to the same error I mentioned, just with `object` in place of `Record<string, unknown>`.

I think maybe I need to set up `TContext` as the type for one of the props? I'm starting to experiment right now, but I'd appreciate any more advice.
